torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1226,9 +1227,12 @@ void ProcessGroupNCCL::heartbeatMonitor() {
   // There might be some cases when we first kill the watchdog thread and get
   // stuck in abort() method, so we want to wait for some buffer time before
   // killing the process.
-  if (terminateProcessGroup_.load() and
+  if ((terminateProcessGroup_.load() || shutDownMode_.load()) &&
       !terminateHeartbeatMonitorThread_.load()) {
-    std::this_thread::sleep_for(std::chrono::seconds(120));
+    auto timeout = heartbeatTimeoutInSec_;
+    // Leave six mins for desync report generation or process group destroy.
+    std::this_thread::sleep_for(
```

Comment:
sigh... 

we're gonna face the same issue Xiaodong mentioned: process-exit is not gonna be fast.  We need to have this sleep also be a CV notifiable one i guess?

Reply:
it might be a simple fix, can you safely use the same CV for this sleep too, and add a .notify call after desync is finished, and also .notify after the ~PGNCCL shutdown sequence finishes and we set the flag to kill watchdog?

Reply:
OK, I think there are couple scenarios:

1. Everything goes super smooth, abort() goes super fast, we send out notify, and then we directly exit in L1201.
2. The thread happens to wake up naturally and there is no heartbeat because watchdog/main thread exit. So we might need a sleep wake up here.
3. We got stuck in abort.

This sleep is trying to differentiate between 2 and 3. Maybe we can the conditional variable wait here, instead of sleep.

Reply:
No, but for case #2, `terminateHeartbeatMonitorThread_` is already set to `true`, there is less likely we will enter this sleep. (I cannot 100% exclude it though) Mostly likely when it hit here, it's either we spend too much time in abort or we are generating desync report.